### PR TITLE
Maintenance Drone Adjustments

### PR DESCRIPTION
**Pull request by Baa14453**
### This PR intends to make Maintenance Drones more practical

- Drones can wire the rooftop solars.
- Drones can carry atmospheric pipe.

### Introduction
So if you've played as a maintenance drone you might have noticed that pulling atmospherics pipes around is a pain, and you can't pull disposal pipe at all (They're size 10!)

### MC introduced
I've added the atmospherics pipe to the magnetic gripper list, this will affect other robots with the gripper, but it's mostly a convenience.

### ~~Epic Battle~~
~~The disposal pipe thing is a pain though, it doesn't really make sense lore-wise allowing drones to pull them since they're big heavy pipes, but it's not very practical game-play wise, and after looking at the code it looks like there was something in place already to allow drones to pull disposal pipe...~~

### ~~The Twist~~
~~It doesn't work though...~~

### ~~The Saviour~~
~~thanks to @Sindorman that's fixed.~~

### Epilogue
Atmospheric pipe doesn't look too good on the hud, not sure if there's a way I can make that better?
Screenshot from 2019-04-13 20-23-13

### Sequel preview
Allows drones walk around the rooftop solars so they can wire them up, I just extended the rooftop solars area to cover the whole walkway.

Before:
roof solars before

After:
rooftop solars after

### Spoilers on twitter
~~Looking for opinions on dragging objects through doors in the same way that drones walk through doors. It'd let them pull things around a bit more seamlessly, but maybe it's a bit OP, maybe if there's a size limit? Would an air pump fit through a drone hatch? idk~~ This part is taking ages so i'll maybe try it in another PR if I can get it to work.

### CODE REVIEW

**Comment by Baa14453** [comment]
!wip

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -675,10 +675,12 @@
 			M.LAssailant = usr
 
 	else if(isobj(AM))
-		var/obj/I = AM
-		if(!can_pull_size || can_pull_size < I.w_class)
-			to_chat(src, "<span class='warning'>It won't budge!</span>")
-			return
+		//Let maintenance drones pull disposal pipe
+		if(!(istype(src, /mob/living/silicon/robot/drone) && istype(AM,/obj/structure/disposalconstruct)))
```
Comment (Sindorman) [inline code/modules/mob/mob.dm]: Terrible, terrible. This is a bad way to make it work. you need to override pulling proc

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 2]
```diff
@@ -327,21 +327,6 @@
 	to_chat(src, "Remember,  you are <b>lawed against interference with the crew</b>. Also remember, <b>you DO NOT take orders from the AI.</b>")
 	to_chat(src, "Use <b>say ;Hello</b> to talk to other drones and <b>say Hello</b> to speak silently to your nearby fellows.")
 
-/mob/living/silicon/robot/drone/start_pulling(var/atom/movable/AM)
```
Comment (Sindorman) [inline code/modules/mob/living/silicon/robot/drone/drone.dm]: Why did you remove this entire proc? All you had to do is remove single check from it.

Comment (Baa14453) [reply code/modules/mob/living/silicon/robot/drone/drone.dm]: I took it out because all the checks it was doing were already checked inside of mob.dm here https://github.com/Aurorastation/Aurora.3/blob/master/code/modules/mob/mob.dm#L656
I didn't realise though that it posts a slightly different message, I guess that's the purpose of it.

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (Baa14453) [reply code/modules/mob/mob.dm]: yeah I thought so, but if I redefine the pulling proc, I think I'd need to redefine the whole thing inside of drone.dm except with that change, because like the previous proc I removed from drone.dm, it'd check if it was a disposal pipe, then go on to execute the parent proc inside mob.dm which wouldn't make the same check.

Comment (Baa14453) [reply code/modules/mob/mob.dm]: What about like this? https://gist.github.com/Baa14453/ecd36628014d3cf875afb02916fcc58d line 330

File: code/modules/mob/living/silicon/robot/drone/drone.dm (hunk 2 above)
Comment (Sindorman) [reply code/modules/mob/living/silicon/robot/drone/drone.dm]: Please put the proc back. it is cleaner this way, I tested it and it works. Here is the code for it.
```
/mob/living/silicon/robot/drone/start_pulling(var/atom/movable/AM)
	if(!AM) return
	for(var/A in pull_list)
		if(istype(AM, A))
			if(pulling)
				var/pulling_old = pulling
				stop_pulling()
				// Are we pulling the same thing twice? Just stop pulling.
				if(pulling_old == AM)
					return

			src.pulling = AM
			AM.pulledby = src

			if(pullin)
				pullin.icon_state = "pull1"
				return
			return
	if(istype(AM, /obj/item))
		var/obj/item/O = AM
		if(O.w_class > can_pull_size)
			to_chat(src, "<span class='warning'>You are too small to pull that.</span>")
			return
	else
		if(!can_pull_mobs)
			to_chat(src, "<span class='warning'>You are too small to pull that.</span>")
			return
	..()
```

Comment (Sindorman) [reply code/modules/mob/living/silicon/robot/drone/drone.dm]: Also add `var/list/pull_list = list(/obj/structure/disposalconstruct)` after line 60

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (Sindorman) [reply code/modules/mob/mob.dm]: With  comment above, revert this proc back to what it was.

File: code/modules/mob/living/silicon/robot/drone/drone.dm (hunk 2 above)
Comment (Baa14453) [reply code/modules/mob/living/silicon/robot/drone/drone.dm]: Ah I see, excellent thanks for your help :D

**Comment by Baa14453** [comment]
!review

**Comment by Baa14453** [comment]
A couple more approval 'ought to do it I reckon @ben10083 
image

**Comment by ben10083** [comment]
0_o I only see one wtf

**Comment by Baa14453** [comment]
> 0_o I only see one wtf

It went away after I refreshed phew...

**Comment by Baa14453** [comment]
> Drones shouldnt be able to move disposal pipes tbh, that should be set back to how it was before.

My argument is that although it makes no sense law-wise, it's impractical and makes it impossible for drones to completely repair a room.
If you look at the original code, it appears to already attempt to override the start_pulling() proc and allow drones to pull the pipe, but it doesn't work.
https://github.com/Aurorastation/Aurora.3/blob/master/code/modules/mob/living/silicon/robot/drone/drone.dm#L330

**Comment by Baa14453** [comment]
> Drones shouldnt be able to move disposal pipes tbh, that should be set back to how it was before.

The subject may be controversial, I'll revert it here and make a separate pull request for it.
